rama-http/src/service/fs/serve_dir/mod.rs

```diff
@@ -487,37 +528,53 @@ enum ServeVariant {
 }
 
 impl ServeVariant {
-    fn build_and_validate_path(&self, base_path: &Path, requested_path: &str) -> Option<PathBuf> {
+    fn build_and_validate_path(&self, source: &DirSource, requested_path: &str) -> Option<PathBuf> {
         match self {
             Self::Directory { serve_mode: _ } => {
                 let path = requested_path.trim_start_matches('/');
 
                 let path_decoded = percent_decode(path.as_ref()).decode_utf8().ok()?;
                 let path_decoded = Path::new(&*path_decoded);
 
-                let mut path_to_file = base_path.to_path_buf();
-                for component in path_decoded.components() {
-                    match component {
-                        Component::Normal(comp) => {
-                            // protect against paths like `/foo/c:/bar/baz` (#204)
-                            if Path::new(&comp)
-                                .components()
-                                .all(|c| matches!(c, Component::Normal(_)))
-                            {
-                                path_to_file.push(comp)
-                            } else {
-                                return None;
+                match source {
+                    DirSource::Filesystem(path) => {
+                        let mut path = path.clone();
+                        for component in path_decoded.components() {
+                            match component {
+                                Component::Normal(comp) => {
+                                    // protect against paths like `/foo/c:/bar/baz` (#204)
+                                    if Path::new(&comp)
+                                        .components()
+                                        .all(|c| matches!(c, Component::Normal(_)))
+                                    {
+                                        path.push(comp)
+                                    } else {
+                                        return None;
+                                    }
+                                }
+                                Component::CurDir => {}
+                                Component::Prefix(_)
+                                | Component::RootDir
+                                | Component::ParentDir => {
+                                    return None;
+                                }
                             }
                         }
-                        Component::CurDir => {}
-                        Component::Prefix(_) | Component::RootDir | Component::ParentDir => {
-                            return None;
-                        }
+                        Some(path)
+                    }
+                    DirSource::Embedded(_) => {
+                        // For embedded directories, return the decoded path
+                        // The actual file lookup will be handled in open_file_embedded
+                        Some(path_decoded.to_path_buf())
```

Comment:
Do we not need similar protection for both sources? Even if it's embedded we are still dealing with same kind of file paths, no?